src/hardware/voodoo.cpp

```diff
@@ -0,0 +1,7175 @@
+	}
+
+	void writeb(PhysPt addr, uint8_t val)
+	{
+		//LOG_MSG("VOODOO: writeb at %x", addr);
+	}
+
+	uint16_t readw(PhysPt addr)
+	{
+		addr = PAGING_GetPhysicalAddress(addr);
+		Bitu retval=voodoo_r((addr>>2)&0x3FFFFF);
+		if (!(addr & 3))
+			retval &= 0xffff;
+		else if (!(addr & 1))
+			retval >>= 16;
+		else
+			E_Exit("voodoo readw unaligned");
+		return retval;
+	}
+
+	void writew(PhysPt addr, uint16_t val)
+	{
+		addr = PAGING_GetPhysicalAddress(addr);
+		if (!(addr & 3))
+			voodoo_w((addr>>2)&0x3FFFFF,(UINT32)val,0x0000ffff);
+		else if (!(addr & 1))
+			voodoo_w((addr>>2)&0x3FFFFF,(UINT32)(val<<16),0xffff0000);
+		else
+			E_Exit("voodoo writew unaligned");
+	}
+
+	uint32_t readd(PhysPt addr)
+	{
+		addr = PAGING_GetPhysicalAddress(addr);
+		if (!(addr&3)) {
+			return voodoo_r((addr>>2)&0x3FFFFF);
+		} else if (!(addr&1)) {
+			Bitu low = voodoo_r((addr>>2)&0x3FFFFF);
+			Bitu high = voodoo_r(((addr>>2)+1)&0x3FFFFF);
+			return (low>>16) | (high<<16);
+		} else {
+			E_Exit("voodoo readd unaligned");
+			return 0xffffffff;
+		}
+	}
+
+	void writed(PhysPt addr, uint32_t val)
+	{
+		addr = PAGING_GetPhysicalAddress(addr);
+		if (!(addr&3)) {
+			voodoo_w((addr>>2)&0x3FFFFF,(UINT32)val,0xffffffff);
+		} else if (!(addr&1)) {
+			voodoo_w((addr>>2)&0x3FFFFF,(UINT32)(val<<16),0xffff0000);
+			voodoo_w(((addr>>2)+1)&0x3FFFFF,(UINT32)val,0x0000ffff);
+		} else {
+			uint32_t val1 = voodoo_r((addr >> 2) & 0x3FFFFF);
+			uint32_t val2 = voodoo_r(((addr >> 2) + 1) & 0x3FFFFF);
+			if ((addr & 3) == 1) {
+				val1 = (val1&0xffffff) | ((val&0xff)<<24);
+				val2 = (val2&0xff000000) | ((UINT32)val>>8);
+			} else if ((addr & 3) == 3) {
+				val1 = (val1&0xff) | ((val&0xffffff)<<8);
+				val2 = (val2&0xffffff00) | ((UINT32)val>>24);
+			}
+			voodoo_w((addr>>2)&0x3FFFFF,val1,0xffffffff);
+			voodoo_w(((addr>>2)+1)&0x3FFFFF,val2,0xffffffff);
+		}
+	}
+} voodoo_real_pagehandler;
+
+static struct Voodoo_Init_PageHandler : public PageHandler {
+	Voodoo_Init_PageHandler() { flags = PFLAG_NOCODE; }
+	uint8_t readb(PhysPt addr)
+	{
+		return 0xff;
+	}
+	uint16_t readw(PhysPt addr)
+	{
+		Voodoo_Startup();
+		return voodoo_real_pagehandler.readw(addr);
+	}
+	uint32_t readd(PhysPt addr)
+	{
+		Voodoo_Startup();
+		return voodoo_real_pagehandler.readd(addr);
+	}
+	void writeb(PhysPt addr, uint8_t val) {}
+	void writew(PhysPt addr, uint16_t val)
+	{
+		Voodoo_Startup();
+		voodoo_real_pagehandler.writew(addr, val);
+	}
+	void writed(PhysPt addr, uint32_t val)
+	{
+		Voodoo_Startup();
+		voodoo_real_pagehandler.writed(addr, val);
+	}
+} voodoo_init_pagehandler;
+
+#define VOODOO_INITIAL_LFB	0xd0000000
+#define VOODOO_REG_PAGES	1024
+#define VOODOO_LFB_PAGES	1024
+#define VOODOO_TEX_PAGES	2048
+#define VOODOO_PAGES (VOODOO_REG_PAGES+VOODOO_LFB_PAGES+VOODOO_TEX_PAGES)
+
+#ifdef C_ENABLE_VOODOO_OPENGL
+#define VOODOO_EMU_TYPE_OFF			0
+#define VOODOO_EMU_TYPE_SOFTWARE	1
+#define VOODOO_EMU_TYPE_ACCELERATED	2
+#endif
+
+static uint32_t voodoo_current_lfb;
```

Comment:
I recommend adding my improvements here:

```suggestion
constexpr uint32_t voodoo_lfb_mask = 0xffff0000;
static uint32_t voodoo_current_lfb = (VOODOO_INITIAL_LFB & voodoo_lfb_mask);

static_assert((VOODOO_INITIAL_LFB & 0xffff0000) == VOODOO_INITIAL_LFB,
              "VOODOO_INITIAL_LFB must have its lower 16 bits set to zero");
```

Reply:
Then remove `#define VOODOO_INITIAL_LFB 0xd0000000`

Reply:
Hmm, I am skipping this improvement for now - I do not understand why here we have different mask here (0xd0000000) compared to `ww/voodoo-3dfx-2` branch (0xffff0000).

@ThomasEricB after this PR will be merged, I suggest try picking these mask-related cleanup (and nothing else!) and create a fresh PR for that. It will be much easier to communicate :)

Reply:
@dreamer It's not having a different mask.... Just for future purposes (since we are doing a new PR for this), 0xffff0000 is actually being called under VOODOO_Init's `voodoo_current_lfb`. It's adding this mask and 0xd0000000 every time it starts, basically.